pkg/ndb/ndb.go

```diff
@@ -0,0 +1,152 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package ndb
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"unsafe"
+
+	dbi "github.com/knqyf263/go-rpmdb/pkg/db"
+	"golang.org/x/xerrors"
+)
+
+type ndbHeader struct {
+	HeaderMagic uint32
+	_           [2]uint32
+	SlotNPages  uint32
+	_           [4]uint32
+}
+
+type ndbSlotEntry struct {
+	SlotMagic uint32
+	PkgIndex  uint32
+	BlkOffset uint32
+	BlkCount  uint32
+}
+
+type ndbBlobHeader struct {
+	BlobMagic uint32
+	PkgIndex  uint32
+	_         uint32
+	BlobLen   uint32
+}
+
+type RpmNDB struct {
+	file  *os.File
+	slots []ndbSlotEntry
+}
+
+const NDB_SlotEntriesPerPage = 4096 / 16 /* 16 == unsafe.Sizeof(NDBSlotEntry) */
+const NDB_HeaderMagic = 'R' | 'p'<<8 | 'm'<<16 | 'P'<<24
+
+func Open(path string) (*RpmNDB, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	hdrBuff := ndbHeader{}
+	err = binary.Read(file, binary.LittleEndian, &hdrBuff)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read metadata: %w", err)
+	}
+
+	if hdrBuff.HeaderMagic != NDB_HeaderMagic || hdrBuff.SlotNPages == 0 {
+		// This does not look like a NDB, return without error
+		return nil, nil
+	}
+
+	// the first two slots are actually the NDB Header
+	slots := make([]ndbSlotEntry, hdrBuff.SlotNPages*NDB_SlotEntriesPerPage-2)
+	err = binary.Read(file, binary.LittleEndian, &slots)
+
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read NDB slots: %w", err)
+	}
+
+	return &RpmNDB{
+		file:  file,
+		slots: slots,
+	}, nil
+}
+
+func (db *RpmNDB) Read() <-chan dbi.Entry {
+	entries := make(chan dbi.Entry)
+
+	go func() {
+		defer close(entries)
+
+		for _, slot := range db.slots {
+			const NDB_SlotMagic = 'S' | 'l'<<8 | 'o'<<16 | 't'<<24
+			if slot.SlotMagic != NDB_SlotMagic {
+				fmt.Println("bad slot magic", slot.SlotMagic)
+				entries <- dbi.Entry{
+					Err: xerrors.Errorf("Bad slot Magic: %x", slot.SlotMagic),
+				}
+				return
+			}
+			// Empty slot?
+			if slot.PkgIndex == 0 {
+				continue
+			}
+			blobHeaderBuff := ndbBlobHeader{}
+			seekOffset := int64(slot.BlkOffset) * int64(unsafe.Sizeof(blobHeaderBuff))
```

Comment:
Do we have to calculate the size of `ndbBlobHeader` every time? Or, can we do it once outside of this function?

Reply:
Ah, I wasn't assuming that this is cost intensive. the size is fixed to 32 bytes, I can just define a constant for it?

Reply:
done.

Reply:
Thanks. It's no big deal actually, but it might be better to avoid doing that many times if possible.